File: android-app/src/main/java/com/medziku/plantwatering/data/parts/RainSensorImpl.java
package com.medziku.plantwatering.data.parts;

import android.os.Parcel;
import android.os.Parcelable;

import com.medziku.plantwatering.station.parts.RainSensor;


public class RainSensorImpl extends RainSensor implements Parcelable {

    public RainSensorImpl(String name, boolean isRaining) {
        super(name, isRaining);
    }

    private RainSensorImpl(Parcel in) {
        super(in.readString(), in.readByte() == 1);
    }

    public static final Creator<RainSensorImpl> CREATOR = new Creator<RainSensorImpl>() {
        @Override
        public RainSensorImpl createFromParcel(Parcel in) {
            return new RainSensorImpl(in);
        }

        @Override
        public RainSensorImpl[] newArray(int size) {
            return new RainSensorImpl[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(getName());
        dest.writeByte(isRaining() ? (byte) 1 : (byte) 0);
    }
}

File: android-app/src/main/java/com/medziku/plantwatering/data/parts/RainSensorImpl.kt
package com.medziku.plantwatering.data.parts

import android.os.Parcel
import android.os.Parcelable

import com.medziku.plantwatering.station.parts.RainSensor


class RainSensorImpl : RainSensor, Parcelable {

    constructor(name: String, isRaining: Boolean) : super(name, isRaining)

    private constructor(`in`: Parcel) : super(`in`.readString(), `in`.readByte().toInt() == 1)

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(name)
        dest.writeByte(if (isRaining) 1.toByte() else 0.toByte())
    }

    companion object {

        val CREATOR: Parcelable.Creator<RainSensorImpl> = object : Parcelable.Creator<RainSensorImpl> {
            override fun createFromParcel(`in`: Parcel): RainSensorImpl {
                return RainSensorImpl(`in`)
            }

            override fun newArray(size: Int): Array<RainSensorImpl?> {
                return arrayOfNulls(size)
            }
        }
    }
}
